### PR TITLE
69 recalculate route

### PR DESCRIPTION
**Pull request by NastyaBondarenko**

### CODE REVIEW

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java [hunk 1]
```diff
@@ -92,4 +122,30 @@ private void processDeliveryStatus(Route route) {
...
+        return routePoints.stream()
+                .allMatch(routePoint -> routePoint.getStatus().equals(PENDING));
+    }
+
+    private void validateDelivery(UUID deliveryId) {
+        Delivery savedDelivery = deliveryRepository.findById(deliveryId).
+                orElseThrow(() -> new DeliveryNotFoundException(deliveryId));
+
+        if (savedDelivery.getStatus().equals(COMPLETED)) {
+            throw new IllegalDeliveryStatusForOperation(savedDelivery, "get");
```
Comment (BogdanParashchak) [inline goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java:136]: non-informative message
throw new IllegalDeliveryStatusForOperation(savedDelivery, "get"); -> throw new IllegalDeliveryStatusForOperation(savedDelivery, "recalculate route for APPROVED delivery ");

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteCalculationService.java [hunk 2]
```diff
@@ -93,6 +93,20 @@ Route mapToRoute(VRPSolution carRoute, StoreDto storeDto) {
         return route;
     }
 
+    @VisibleForTesting
+    Route mapToRoute(Route route, LinkedList<RoutePoint> routePoints) {
```
Comment (airavata-elephant) [inline goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteCalculationService.java:97]: maybe a better name for the method is: "recalculateRoute" ?

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java [hunk 3]
```diff
@@ -92,4 +122,30 @@ private void processDeliveryStatus(Route route) {
...
+
+    private void validateRoute(Route route, UUID deliveryId) {
+        if (!route.getDelivery().getId().equals(deliveryId)) {
+            throw new RouteNotFoundException(deliveryId);
+        }
+        RouteStatus routeStatus = route.getStatus();
+        if (routeStatus.equals(RouteStatus.COMPLETED) ||
+                routeStatus.equals(RouteStatus.INPROGRESS) ||
+                routeStatus.equals(RouteStatus.INCOMPLETE)) {
+            throw new IllegalRouteStatusForOperation(route, routeStatus.toString());
```
Comment (BogdanParashchak) [inline goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java:148]: looks like mistake, see:
    public IllegalRouteStatusForOperation(Route route, String action) {
        super(String.format(ILLEGAL_STATUS_MESSAGE, action, route.getId(), route.getStatus()));
    }

**Review by airavata-elephant — COMMENTED** [review]
check comment plz

**Review by BogdanParashchak — CHANGES_REQUESTED** [review]
see comments

File: goods-partner-backend/src/main/java/com/goodspartner/web/controller/DeliveryController.java [hunk 4]
```diff
@@ -128,4 +129,11 @@ public void updateRoute(@PathVariable int routeId, @RequestBody RouteDto routeDt
     public void updateRoutePoint(@PathVariable int routeId, @PathVariable String routePointId, @RequestBody RoutePoint routePoint) {
         routeService.updatePoint(routeId, routePointId, routePoint);
     }
+
+    @PreAuthorize("hasAnyRole('ADMIN', 'LOGIST', 'DRIVER')")
+    @PutMapping("/{id}/routes/{routeId}/recalculate")
+    public void updateRecalculatedRoute(@PathVariable("id") UUID deliveryId, @PathVariable int routeId,
```
Comment (belosh59) [inline goods-partner-backend/src/main/java/com/goodspartner/web/controller/DeliveryController.java:135]: I would call it /reorder coz in the end we dont calculate the route we reordering it

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteCalculationService.java (hunk 2 above)
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteCalculationService.java]: +

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java [hunk 5]
```diff
@@ -69,6 +80,24 @@ public void updatePoint(int routeId, String routePointId, RoutePoint routePoint)
         routeRepository.save(route);
     }
 
+    @Override
+    public void updateRoute(UUID deliveryId, int routeId, LinkedList<RoutePoint> routePoints) {
+
+        validateDelivery(deliveryId);
+
+        Route savedRoute = routeRepository.findById(routeId)
```
[the remark this answers is not in the document]
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java]: changed

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java [hunk 6]
```diff
@@ -69,6 +80,24 @@ public void updatePoint(int routeId, String routePointId, RoutePoint routePoint)
...
+
+        validateDelivery(deliveryId);
+
+        Route savedRoute = routeRepository.findById(routeId)
+                .orElseThrow(() -> new RouteNotFoundException(routeId));
+
+        validateRoute(savedRoute, deliveryId);
+
+        if (isAllRoutePointsPending(routePoints)) {
+            Route updatedRoute = routeCalculationService.mapToRoute(savedRoute, routePoints);
```
[the remark this answers is not in the document]
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java]: changed

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java (hunk 1 above)
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java]: changed

File: goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java (hunk 3 above)
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/service/impl/DefaultRouteService.java]: changed

File: goods-partner-backend/src/main/java/com/goodspartner/web/controller/DeliveryController.java (hunk 4 above)
Comment (NastyaBondarenko) [reply goods-partner-backend/src/main/java/com/goodspartner/web/controller/DeliveryController.java]: changed

**Review by belosh59 — APPROVED** [review]

**Review by belosh59 — APPROVED** [review]

**Review by BogdanParashchak — APPROVED** [review]

**Review by belosh59 — CHANGES_REQUESTED** [review]
need time re-evaluation

**Review by belosh59 — APPROVED** [review]

**Pull request merged**